contracts/plugins/assets/morpho-aave/MorphoAaveV2TokenisedDeposit.sol

```diff
@@ -0,0 +1,120 @@
+            config.rewardToken
+        )
+    {
+        underlying = address(config.underlyingERC20);
+        morphoController = config.morphoController;
+        morphoLens = config.morphoLens;
+        poolToken = address(config.poolToken);
+        rewardsDistributor = config.rewardsDistributor;
+    }
+
+    // Offchain keepers should periodically call this such that we accumuate rewards
+    function claimOffchainRewards(uint256 _claimable, bytes32[] calldata _proof) external {
+        uint256 delta;
+        uint256 _totalSupply = totalSupply();
+        if (_totalSupply > 0) {
+            uint256 initialBal = rewardToken.balanceOf(address(this));
+            rewardsDistributor.claim(address(this), _claimable, _proof);
+            uint256 endingBal = rewardToken.balanceOf(address(this));
+            delta = endingBal - initialBal;
+
+            // {qRewards/share} += {qRewards} * {qShare/share} / {qShare}
+            rewardsPerShare += ((delta) * one) / _totalSupply;
+        }
+        _syncAccount(msg.sender);
+        _claimAccountRewards(msg.sender);
+    }
+
+    // solhint-disable-next-line no-empty-blocks
+    function _claimAssetRewards() internal virtual override {}
+
+    function totalAssets() public view virtual override returns (uint256) {
+        (, , uint256 currentBalance) = morphoLens.getCurrentSupplyBalanceInOf(
+            poolToken,
+            address(this)
+        );
+        return currentBalance;
+    }
+
+    function _deposit(
+        address caller,
+        address receiver,
+        uint256 assets,
+        uint256 shares
+    ) internal virtual override {
+        SafeERC20.safeTransferFrom(IERC20(underlying), caller, address(this), assets);
+        SafeERC20.safeApprove(IERC20(underlying), address(morphoController), assets);
+        morphoController.supply(poolToken, assets);
+
+        _mint(receiver, shares);
+        emit Deposit(caller, receiver, assets, shares);
+    }
+
+    function _decimalsOffset() internal view virtual override returns (uint8) {
+        return 0;
+    }
+
+    function _withdraw(
+        address caller,
+        address receiver,
+        address owner,
+        uint256 assets,
+        uint256 shares
+    ) internal virtual override {
+        if (caller != owner) {
+            _spendAllowance(owner, caller, shares);
+        }
+        morphoController.withdraw(poolToken, assets);
+
+        _burn(owner, shares);
+        emit Withdraw(caller, receiver, owner, assets, shares);
+    }
+
+    function getExchangeRate() external view virtual override returns (uint192) {
+        (, , uint256 currentBalance) = morphoLens.getCurrentSupplyBalanceInOf(
+            poolToken,
+            address(this)
+        );
+        if (currentBalance == 0) {
+            return FIX_ONE;
+        }
+        return
+            shiftl_toFix(
+                _convertToShares(10**ERC20(underlying).decimals(), Math.Rounding.Down),
```

Comment:
minor optimization: I think it's worth making `ERC20(underlying).decimals()` an immutable var just given how many times this exchange rate is going to be computed